File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
def search_first(nums: list[int], target: int) -> int:
    low = 0
    high = len(nums) - 1
    while low <= high:
        mid = (low + high) // 2
        if nums[mid] == target:
            if mid != 0 and nums[mid-1] != target:
                return mid
            elif mid == 0:
                return mid
            else:
                high = mid - 1
        elif nums[mid] < target:
            low = mid + 1
        elif nums[mid] > target:
            high = mid - 1
    return -1

def search_last(nums: list[int], target: int) -> int:
    low = 0
    high = len(nums) - 1
    while low <= high:
        mid = (low + high) // 2
        if nums[mid] == target:
            if mid != len(nums) - 1 and nums[mid+1] != target:
                return mid
            elif mid == len(nums) - 1:
                return mid
            else:
                low = mid + 1
        elif nums[mid] < target:
            low = mid + 1
        elif nums[mid] > target:
            high = mid - 1
    return -1

class Solution:
    
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        first = search_first(nums, target)
        if first == -1:
            return [-1, -1]
        last = search_last(nums, target)
        ans = [first, last]
        return ans

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.cpp
#include <vector>

class Solution {
private:
    int search_first(const std::vector<int>& nums, int target) {
        int low = 0;
        int high = static_cast<int>(nums.size()) - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (nums[mid] == target) {
                if (mid != 0 && nums[mid - 1] != target) {
                    return mid;
                } else if (mid == 0) {
                    return mid;
                } else {
                    high = mid - 1;
                }
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }

    int search_last(const std::vector<int>& nums, int target) {
        int low = 0;
        int high = static_cast<int>(nums.size()) - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (nums[mid] == target) {
                if (mid != static_cast<int>(nums.size()) - 1 && nums[mid + 1] != target) {
                    return mid;
                } else if (mid == static_cast<int>(nums.size()) - 1) {
                    return mid;
                } else {
                    low = mid + 1;
                }
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }

public:
    std::vector<int> searchRange(std::vector<int>& nums, int target) {
        int first = search_first(nums, target);
        if (first == -1) {
            return {-1, -1};
        }
        int last = search_last(nums, target);
        return {first, last};
    }
};